irohad/ametsuchi/impl/postgres_command_executor.cpp

```diff
@@ -1197,6 +1224,44 @@ namespace iroha {
           sql_, cmd.str(), "TransferAsset", std::move(str_args));
     }
 
+    CommandResult PostgresCommandExecutor::operator()(
+        const shared_model::interface::CompareAndSetAccountDetail &command) {
+      auto &account_id = command.accountId();
+      auto &key = command.key();
+      auto &value = command.value();
+      auto &old_value = command.oldValue();
+
+      if (creator_account_id_.empty()) {
+        // When creator is not known, it is genesis block
+        creator_account_id_ = "genesis";
+      }
+      std::string val = "\"" + value + "\"";
+      std::string old_val = "\"\"";
+      if (old_value)
+      {
+        old_val = "\"" + *old_value + "\"";
+
+      }
+
+      auto cmd = boost::format("EXECUTE %1% ('%2%', '%3%', '%4%', '%5%', '%6%')");
+
+      appendCommandName("compareAndSetAccountDetail", cmd, do_validation_);
+
+      cmd = (cmd % creator_account_id_ % account_id % key % val % old_val);
+
+      auto str_args = [&account_id, &key, &value, &old_value] {
+        return getQueryArgsStringBuilder()
+            .append("account_id", account_id)
+            .append("key", key)
+            .append("value", value)
+            .append("old_value", old_value ? *old_value : "")
```

Comment:
I think, that it is not a correct way
```expected_json_value``` = "```old_value```"
My suggestion is ```command.oldValue().value_or("")```

Reply:
``` auto old_value = command.oldValue().value_or("");```
```std::string expected_json_value = "\"" + old_value + "\"";```
```            .append("old_value", old_value)```

Reply:
looks good, but please take a look at my main review message

Reply:
I think you are right about the difference between empty string and NULL
So my solution is :
1. Removed single quotes from sixth parametr 
```C++ 
auto cmd = boost::format("EXECUTE %1% ('%2%', '%3%', '%4%', '%5%', %6%)");
``` 
It is need for pass NULL as a parametr
2. By default, old_value NULL, but if command have it, then we wrap it into single and double quotes
```C++
      std::string expected_json_value = "NULL";`

      if (old_value) {
        expected_json_value = "'\"" + old_value.get() + "\"'";
      }
```
3. Function old_value now have to checks 
  1) That these key-value pair exist
  1.1) If it exists then checks that expected_value is not NULL, otherwise just create a new pair
  1.1.1) If it is not NULL then compare with old value, otherwise expected_value didn't present
```SQL
old_value AS
(
  SELECT *
  FROM account
  WHERE
    account_id = $2
    AND CASE
      WHEN data ? $1 AND data->$1 ?$3
        THEN
          CASE
            WHEN $5 IS NOT NULL THEN data->$1->$3 = $5::jsonb
              ELSE FALSE
          END
        ELSE TRUE
      END
)
```

Reply:
nice. you can also benefit from boost::optional integration in soci: boost::none translates to NULL automatically, so you do not have to write it yourself

Reply:
Maybe, but I don't know-how

Reply:
> you can also benefit from boost::optional integration in soci: boost::none translates to NULL automatically, so you do not have to write it yourself

Fixed, in my way, but I would be interested to know about your method

Reply:
seems I have mislead you: we have no convenient way to bind arguments with `soci::use` in command executor, like we do in query executor. 


>     1. Removed single quotes from sixth parametr

and OMG!! I just read the executor code and see a huge flaw SQLI :open_mouth: 
It was there before your changes, do not worry

Reply:
> huge flaw SQLI
SQL injection?

Reply:
yep. fixing it right now